### PR TITLE
Improvement/bb 541 avoid initializing mongodb client

### PR DESCRIPTION
**Pull request by benzekrimaha**
Issue : BB-541

### CODE REVIEW

File: lib/api/BackbeatAPI.js [hunk 1]
```diff
@@ -1260,40 +1260,12 @@ class BackbeatAPI {
...
-                    this._logger.error('error setting up location status manager', {
-                        method: 'BackbeatAPI.setupInternals',
-                        error: err.message,
-                    });
-                    return cb(err);
-                }
-                this._logger.info('BackbeatAPI setup ready');
+            if (this._config.queuePopulator.mongo) {
+                return this._setupLocationStatusManager(cb);
+            } else {
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]: we still need to initialize the `LocationStatusManager` as the pause/resume of crr uses redis/zookeeper and not mongodb and so we need to keep it. This case should be handled inside the `LocationStatusManager` class.

Comment (welansari) [reply lib/api/BackbeatAPI.js]: Actually, pause/resume was not available on S3C, maybe we should keep as is for now without adding new features.

File: lib/api/BackbeatAPI.js [hunk 2]
```diff
@@ -1335,28 +1307,73 @@ class BackbeatAPI {
...
+
+        return this._locationStatusManager._setupLocationStatusStore(err => {
+            if (err) {
+                this._logger.error('error setting up location status manager', {
+                    method: 'BackbeatAPI.setupInternals',
+                    error: err.message,
+                });
+                return cb(err);
+            }
+            this._logger.info('BackbeatAPI setup ready');
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]: Does it really make sense to keep this log inside this function ?

File: lib/api/BackbeatAPI.js [hunk 3]
```diff
@@ -1335,28 +1307,73 @@ class BackbeatAPI {
...
+                    isMongo: true,
+                }
+            },
+            this._logger,
+        );
+
+        return this._locationStatusManager._setupLocationStatusStore(err => {
+            if (err) {
+                this._logger.error('error setting up location status manager', {
+                    method: 'BackbeatAPI.setupInternals',
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]:
```suggestion
                    method: 'BackbeatAPI._setupLocationStatusManager',
```

File: lib/api/BackbeatAPI.js [hunk 4]
```diff
@@ -1335,28 +1307,73 @@ class BackbeatAPI {
      * @returns {undefined} undefined
      */
     _setupMongoClient(cb) {
-        const mongoConfig = this._config.queuePopulator.mongo;
-        const mongoUrl = constructConnectionString(mongoConfig);
-        MongoClient.connect(mongoUrl, {
+        if (this._config.queuePopulator.mongo) {
+            const mongoConfig = this._config.queuePopulator.mongo;
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]:
```suggestion
        const mongoConfig = this._config.queuePopulator.mongo;
        if (!mongoConfig) {
            this._logger.debug('MongoDB configuration not found, skipping MongoDB client setup', {
                method: 'BackbeatAPI._setupMongoClient',
            });
            return cb();
         }
```
Just my opinion, but i think doing it the other way is cleaner ?

File: lib/api/BackbeatAPI.js [hunk 5]
```diff
@@ -1260,40 +1260,12 @@ class BackbeatAPI {
...
-                        method: 'BackbeatAPI.setupInternals',
-                        error: err.message,
-                    });
-                    return cb(err);
-                }
-                this._logger.info('BackbeatAPI setup ready');
+            if (this._config.queuePopulator.mongo) {
+                return this._setupLocationStatusManager(cb);
+            } else {
+                this._logger.debug('BackbeatAPI setup ready without Mongo');
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]: Redundant with the log in `_setupMongoClient`

**Review by KillianG — APPROVED** [review]

File: lib/api/BackbeatAPI.js [hunk 6]
```diff
@@ -1260,40 +1260,11 @@ class BackbeatAPI {
...
-            return this._locationStatusManager._setupLocationStatusStore(err => {
-                if (err) {
-                    this._logger.error('error setting up location status manager', {
-                        method: 'BackbeatAPI.setupInternals',
-                        error: err.message,
-                    });
-                    return cb(err);
-                }
-                this._logger.info('BackbeatAPI setup ready');
+            if (this._config.queuePopulator.mongo) {
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]:
```suggestion
            if (this._queuePopulator.mongo) {
```

File: lib/api/BackbeatAPI.js [hunk 7]
```diff
@@ -1260,40 +1260,11 @@ class BackbeatAPI {
...
+                return this._setupLocationStatusManager(cb);
+            } else {
                 return cb();
-            });
+            }
```
Comment (welansari) [inline lib/api/BackbeatAPI.js]:
```suggestion
                return this._setupLocationStatusManager(cb);
            }
            return cb();
```

**Review by welansari — APPROVED** [review]

File: lib/api/BackbeatAPI.js [hunk 8]
```diff
@@ -1336,26 +1307,70 @@ class BackbeatAPI {
...
+            this._logger.debug('MongoDB configuration not found, skipping MongoDB client setup', {
+                method: 'BackbeatAPI._setupMongoClient',
+            });
+            return cb();
+        } else {
+            const mongoUrl = constructConnectionString(mongoConfig);
+            return MongoClient.connect(mongoUrl, {
                 replicaSet: mongoConfig.replicaSet,
                 useNewUrlParser: true,
                 useUnifiedTopology: true,
```
Comment (ghost) [inline lib/api/BackbeatAPI.js:1321]: Not related to the changes, but I believe this is not used anymore with recent mongodb drivers, but I also see that we are still using mongo driver v3 in Backbeat... Would it make sense to bump it in this PR, as we modify this function? or we can do it when upgrading the deps (but these lines will be easy to miss...)

Comment (benzekrimaha) [reply lib/api/BackbeatAPI.js:1321]: I suggest adding a comment so it is not missed during the dependencies bump

**Pull request merged**